src/waveform/renderers/allshader/waveformrenderersimple.cpp

```diff
@@ -1,172 +1,162 @@
-    m_vertices[1].reserve(reserved[1]);
+    geometry().setDrawingMode(Geometry::DrawingMode::Triangles);
+    geometry().allocate(reserved);
+    markDirtyGeometry();
 
-    m_vertices[1].addRectangle(
-            0.f,
-            halfBreadth - 0.5f * devicePixelRatio,
-            static_cast<float>(length),
-            halfBreadth + 0.5f * devicePixelRatio);
+    RGBVertexUpdater vertexUpdater{geometry().vertexDataAs<Geometry::RGBColoredPoint2D>()};
+    vertexUpdater.addRectangle({0.f,
+                                       halfBreadth - 0.5f},
+            {static_cast<float>(length),
+                    halfBreadth + 0.5f},
+            {static_cast<float>(m_axesColor_r),
+                    static_cast<float>(m_axesColor_g),
+                    static_cast<float>(m_axesColor_b)});
 
-    // We will iterate over a range of waveform data, centered around xVisualFrame
     const double maxSamplingRange = visualIncrementPerPixel / 2.0;
 
-    for (int pos = 0; pos < length; ++pos) {
-        // Calculate the start and end of the range of waveform data, centered around xVisualFrame
+    const QVector3D signalColor{static_cast<float>(m_signalColor_r),
+            static_cast<float>(m_signalColor_g),
+            static_cast<float>(m_signalColor_b)};
+
+    for (int pos = 0; pos < pixelLength; ++pos) {
         const int visualFrameStart = std::lround(xVisualFrame - maxSamplingRange);
         const int visualFrameStop = std::lround(xVisualFrame + maxSamplingRange);
 
-        // Calculate the actual (deinterleaved) indices.
-        //
-        // Make sure we stay inside data at the lower boundary
         const int visualIndexStart = std::max(visualFrameStart * 2, 0);
-        // and at the upper boundary.
-        // Note: * dataSize - 1, because below we add chn = 1
-        //       * visualFrameStart + 1, because we want to have at least 1 value
         const int visualIndexStop =
                 std::min(std::max(visualFrameStop, visualFrameStart + 1) * 2, dataSize - 1);
 
-        // 2 channels
-        float max[2]{};
+        const float fpos = static_cast<float>(pos) * invDevicePixelRatio;
 
+        // - Per channel
+        uchar u8maxAllChn[2]{};
         for (int chn = 0; chn < 2; chn++) {
             // data is interleaved left / right
             for (int i = visualIndexStart + chn; i < visualIndexStop + chn; i += 2) {
                 const WaveformData& waveformData = data[i];
-                const float filteredAll = static_cast<float>(waveformData.filtered.all);
 
-                max[chn] = math_max(max[chn], filteredAll);
+                u8maxAllChn[chn] = math_max(u8maxAllChn[chn], waveformData.filtered.all);
             }
         }
+        float maxAllChn[2]{static_cast<float>(u8maxAllChn[0]), static_cast<float>(u8maxAllChn[1])};
 
-        const float fpos = static_cast<float>(pos);
+        // TODO: use two geometrynodes, with uniform material,
+        // one for the axis, one for the signal
```

Comment:
Is this `TODO` still outstanding as part of this PR?

Reply:
I forgot about it, but since it's really just a minor optimisation, I think it doesn't block this PR.